src/Microsoft.Diagnostics.Monitoring.WebApi/EventMonitoringPassthroughStream.cs

```diff
@@ -0,0 +1,256 @@
+        /// <param name="onEvent">A callback that will be invoked each time the requested event has been observed.</param>
+        /// <param name="onPayloadFilterMismatch">A callback that will be invoked if the field names specified in <paramref name="payloadFilter"/> do not match those in the event's manifest.</param>
+        /// <param name="sourceStream">The source event stream which is compatible with <see cref="EventPipeEventSource"/>.</param>
+        /// <param name="destinationStream">The destination stream to write events. It must either be full duplex or be write-only.</param>
+        /// <param name="bufferSize">The size of the buffer to use when writing to the <paramref name="destinationStream"/>.</param>
+        /// <param name="callOnEventOnlyOnce">If true, the provided <paramref name="onEvent"/> will only be called for the first matching event.</param>
+        /// <param name="leaveDestinationStreamOpen">If true, the provided <paramref name="destinationStream"/> will not be automatically closed when this class is.</param>
+        public EventMonitoringPassthroughStream(
+            string providerName,
+            string eventName,
+            IDictionary<string, string> payloadFilter,
+            Action<TraceEvent> onEvent,
+            Action<TraceEvent> onPayloadFilterMismatch,
+            Stream sourceStream,
+            Stream destinationStream,
+            int bufferSize,
+            bool callOnEventOnlyOnce,
+            bool leaveDestinationStreamOpen) : base()
+        {
+            _providerName = providerName;
+            _eventName = eventName;
+            _onEvent = onEvent;
+            _onPayloadFilterMismatch = onPayloadFilterMismatch;
+            _sourceStream = sourceStream;
+            _payloadFilter = payloadFilter;
+            _callOnEventOnlyOnce = callOnEventOnlyOnce;
+
+            // Wrap a buffered stream around the destination stream
+            // to avoid slowing down the event processing with the data
+            // passthrough unless there is significant pressure.
+            _destinationStream = new BufferedStream(
+                leaveDestinationStreamOpen
+                    ? new StreamLeaveOpenWrapper(destinationStream)
+                    : destinationStream,
+                bufferSize);
+        }
+
+        /// <summary>
+        /// Start processing the event stream, monitoring it for the requested event and transferring its data to the specified destination stream.
+        /// This will continue to run until the event stream is complete or a stop is requested, regardless of if the requested event has been observed.
+        /// </summary>
+        /// <param name="token">The cancellation token. It can only be signaled before processing has been started. After that point <see cref="DisposeAsync"/> should be called to stop processing.</param>
```

Comment:
Can you register a callback on the cancellation token within the `Task.Run` anonymous function that disposes the event source and then call `token.ThrowIfCancellationRequested` after the call to `_eventSource.Process` returns? It looks like `EventPipeEventSource ` disposal is idempotent. This would then match the expected semantics for async methods.

Reply:
You may want to add a `token.ThrowIfCancellationRequested` call after the construction of the `EventPipeEventSource` as well. The ctor frequently hangs for several seconds waiting for some events from the runtime.